### PR TITLE
Feature: ios v4 support

### PR DESCRIPTION
**Pull request by aboedo**
### Requirements: 
- [x] ~purchases-hybrid-common release with iOS v4 APIs~

### Description: 

This PR updates the internal usage of iOS classes to the latest purchases-ios / purchases-hybrid-common. It doesn't update the public APIs exposed by purchases-unity, though - this will be a separate PR

### Notes: 

This PR looks a lot bigger than it really is - most of it is just unity meta files, autogenerated and safe to ignore. It's actually just a couple hundred lines

### CODE REVIEW

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 1]
```diff
@@ -128,16 +128,16 @@ - (void)purchasePackage:(NSString *)packageIdentifier
...
-    [RCCommonFunctionality restoreTransactionsWithCompletionBlock:[self getPurchaserInfoCompletionBlockFor:RESTORE_TRANSACTIONS]];
+- (void)restorePurchases {
+    [RCCommonFunctionality restorePurchasesWithCompletionBlock:[self getCustomerInfoCompletionBlockFor:RESTORE_TRANSACTIONS]];
 }
 
 - (void)syncPurchases {
     // on Android, syncPurchases doesn't have a completion block. So instead of
-    // calling getPurchaserInfoCompletionBlockFor:SYNC_PURCHASES, we just
+    // calling getCustomerInfoCompletionBlockFor:SYNC_PURCHASES, we just
     // print the response, to match Android behavior.
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:138]: I guess we missed this inconsistency in android v5 / iOS v4 😬

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 2]
```diff
@@ -226,7 +214,7 @@ - (char *)getAppUserID {
 }
 
 - (BOOL)isAnonymous {
-    return @([RCCommonFunctionality isAnonymous]);
+    return RCCommonFunctionality.isAnonymous;
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:217]: this was a bug - doing the `@([` on a boolean always evaluates to true

File: RevenueCat/Scripts/PurchasesWrapperiOS.cs [hunk 3]
```diff
@@ -361,10 +355,10 @@ public void CanMakePayments(Purchases.BillingFeature[] features)
...
     [DllImport("__Internal")]
-    private static extern void _RCGetPaymentDiscount(string productIdentifier, string discountIdentifier);
+    private static extern void _RCGetPromotionalOffer(string productIdentifier, string discountIdentifier);
     public void GetPaymentDiscount(string productIdentifier, string discountIdentifier)
     {
-        _RCGetPaymentDiscount(productIdentifier, discountIdentifier);
+        _RCGetPromotionalOffer(productIdentifier, discountIdentifier);
     }
     
 }
```
Comment (aboedo) [inline RevenueCat/Scripts/PurchasesWrapperiOS.cs:364]: from this point on, the rest of the changes are unity automated changes

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 4]
```diff
@@ -19,7 +20,7 @@
 static NSString *const RESET = @"_reset";
 static NSString *const MAKE_PURCHASE = @"_makePurchase";
 static NSString *const GET_OFFERINGS = @"_getOfferings";
-static NSString *const GET_PURCHASER_INFO = @"_getPurchaserInfo";
+static NSString *const GET_PURCHASER_INFO = @"_getCustomerInfo";
```
Comment (vegaro) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:23]:
```suggestion
static NSString *const GET_CUSTOMER_INFO = @"_getCustomerInfo";
```

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 5]
```diff
@@ -65,17 +66,16 @@ - (void)setupPurchases:(NSString *)apiKey
...
     self.gameObject = nil;
 
-    [RCPurchases configureWithAPIKey:apiKey
-                           appUserID:appUserID
-                        observerMode:observerMode
-               userDefaultsSuiteName:userDefaultsSuiteName
-                      platformFlavor:self.platformFlavor
-               platformFlavorVersion:self.platformFlavorVersion
-                   dangerousSettings:nil];
+    RCPurchases *instance __unused = [RCPurchases configureWithAPIKey:apiKey
```
Comment (vegaro) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:69]: Why are you storing the instance?

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m (hunk 2 above)
Comment (vegaro) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:217]: 😮‍💨

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 6]
```diff
@@ -263,16 +251,16 @@ - (void)canMakePaymentsWithFeatures:(NSArray<NSNumber *> *)features {
...
+                                    discount:(NSString *)discountIdentifier {
+    [RCCommonFunctionality promotionalOfferForProductIdentifier:productIdentifier
                                                       discount:discountIdentifier
                                                completionBlock:^(NSDictionary *_Nullable responseDictionary, RCErrorContainer *_Nullable error) {
         [self sendJSONObject:responseDictionary toMethod:GET_PAYMENT_DISCOUNT];
     }];
 }
 
-#pragma mark - Subcriber Attributes
+#pragma mark - Susbcriber Attributes
```
Comment (vegaro) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:263]: Lol, there's still a typo


```suggestion
#pragma mark - Subscriber Attributes
```

File: RevenueCat/Scripts/PurchasesWrapperiOS.cs [hunk 7]
```diff
@@ -93,25 +93,19 @@ public void LogOut()
         _RCLogOut();
     }
 
-    [DllImport("__Internal")]
-    private static extern void _RCCreateAlias(string newAppUserId);
     public void CreateAlias(string newAppUserId)
     {
-        _RCCreateAlias(newAppUserId);
+        // todo: no-op for now but this method should be removed
```
Comment (vegaro) [inline RevenueCat/Scripts/PurchasesWrapperiOS.cs:98]: can we remove it?

File: Subtester/ProjectSettings/ProjectVersion.txt [hunk 8]
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 2019.4.13f1
-m_EditorVersionWithRevision: 2019.4.13f1 (518737b1de84)
+m_EditorVersion: 2021.2.13f1
```
Comment (vegaro) [inline Subtester/ProjectSettings/ProjectVersion.txt:1]: oh nice you updated the Unity version! Great I've been wanting to do this for a while haha. Do you think we have to specify which version of Unity we support (meaning the one we have tested with)? I think we should match Unity IAPs min version

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m (hunk 4 above)
Comment (aboedo) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:23]: 👍

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m (hunk 5 above)
Comment (aboedo) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:69]: I genuinely don't remember why I did that ¯\\_(ツ)_/¯ removing

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m (hunk 6 above)
Comment (aboedo) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:263]: 😂

File: RevenueCat/Scripts/PurchasesWrapperiOS.cs (hunk 7 above)
Comment (aboedo) [reply RevenueCat/Scripts/PurchasesWrapperiOS.cs:98]: I'm gonna do it in the next PR, this one just maps existing APIs to v4

File: Subtester/ProjectSettings/ProjectVersion.txt (hunk 8 above)
Comment (aboedo) [reply Subtester/ProjectSettings/ProjectVersion.txt:1]: That's a great idea! 
They're currently releasing for 2021.3 https://docs.unity3d.com/Manual/com.unity.purchasing.html. 
I'll update to that one in a separate PR

**Comment by aboedo** [comment]
@vegaro this is ready for another pass

**Review by vegaro — APPROVED** [review]

**Pull request merged**